packages/libs/react-ui/src/components/Button/Button.tsx

```diff
@@ -1,87 +1,264 @@
 import { mergeProps, useObjectRef } from '@react-aria/utils';
+import type { RecipeVariants } from '@vanilla-extract/recipes';
 import classNames from 'classnames';
-import type { ForwardedRef } from 'react';
-import React, { forwardRef } from 'react';
-import type { AriaButtonProps } from 'react-aria';
-import { useButton, useFocusRing, useHover } from 'react-aria';
+import type {
+  ComponentProps,
+  ForwardedRef,
+  MouseEventHandler,
+  ReactElement,
+  ReactNode,
+} from 'react';
+import React, { cloneElement, forwardRef } from 'react';
+import type { AriaButtonProps, AriaFocusRingProps } from 'react-aria';
+import { useButton, useFocusRing, useHover, useLink } from 'react-aria';
+import type { IAvatarProps } from '../Avatar';
+import { Avatar } from '../Avatar';
+import { Badge } from '../Badge';
 import { ProgressCircle } from '../ProgressCircle/ProgressCircle';
-import { button } from './SharedButton.css';
-import type { ISharedButtonProps } from './utils';
+import {
+  avatarStyle,
+  badgeStyle,
+  button,
+  centerContentWrapper,
+  directionStyle,
+  disabledBadgeStyle,
+  iconOnlyStyle,
+  iconStyle,
+  noPostfixStyle,
+  noPrefixStyle,
+  postfixIconStyle,
+  prefixIconStyle,
+} from './Button.css';
 import { disableLoadingProps } from './utils';
 
-// omit link related props from `AriaButtonProps`
-type PickedAriaButtonProps = Omit<
-  AriaButtonProps,
-  'href' | 'target' | 'rel' | 'elementType'
->;
+type Variants = NonNullable<RecipeVariants<typeof button>>;
 
-export interface IButtonProps
-  extends PickedAriaButtonProps,
-    ISharedButtonProps {}
+type BaseProps = Omit<AriaFocusRingProps, 'isTextInput'> & Variants;
+
+export interface ICustomProps extends BaseProps {
+  avatarProps?: Omit<IAvatarProps, 'size'>;
+  badgeValue?: string | number;
```

Comment:
I think that either way we are bound to that. (my thoughts) If you would provide it as a child component and there is a breaking change you still need to change it. One way to mitigate it could be to split styling and functionality. Set up a hook for the styling and copy the component to prevent API changes from breaking but still allow consistency with styling. 
This way you have less components to replace on a breaking change in one go but don't have feature parity with the actual component and will probably have the breaking change on a later stage in the button if needed. 

Let's have a discussion about this during refinement, I would love to explore other approaches.